src/tsWorker.ts

```diff
@@ -50,6 +50,255 @@ export enum CodeOutlineTokenKind {
 	Set = 'Set'
 }
 
+
+/**
+ * An enum containing constants that describe the kind of each class member.
+ */
+ enum ThingworxTypescriptWidgetClassMemberKind {
+	/**
+	 * Indicates that this member represents a widget property.
+	 */
+	Property = 'property',
+
+	/**
+	 * Indicates that this member represents a widget event.
+	 */
+	Event = 'event',
+
+	/**
+	 * Indicates that this member represents a widget service.
+	 */
+	Service = 'service'
+}
+
+/**
+ * An interface describing a widget class member aspect.
+ */
+interface ThingworxTypescriptWidgetAspect {
+	/**
+	 * The name of the aspect.
+	 */
+	name: string;
+
+	/**
+	 * An optional argument that the aspect has been invoked with, if this aspect is callable.
+	 */
+	value?: unknown;
+}
+
+/**
+ * An interface describing a widget class member.
+ */
+interface ThingworxTypescriptWidgetClassMember {
+	/**
+	 * The name of this class member.
+	 */
+	name: string;
+
+	/**
+	 * An optional list of aspects that may be associated with this member.
+	 */
+	aspects?: ThingworxTypescriptWidgetAspect[];
+
+	/**
+	 * An optional description to display for this member in the composer.
+	 */
+	description?: string;
+
+	/**
+	 * The kind of widget member represented by this class member.
+	 */
+	kind: ThingworxTypescriptWidgetClassMemberKind;
+}
+
+interface ThingworxTypescriptWidgetProperty extends ThingworxTypescriptWidgetClassMember {
+	/**
+	 * The property's base type.
+	 */
+	baseType: string;
+
+	kind: ThingworxTypescriptWidgetClassMemberKind.Property;
+}
+
+/**
+ * An interface that represents the definition of a widget class.
+ */
+interface ThingworxTypescriptWidgetClass {
+	/**
+	 * The name of the widget class.
+	 */
+	name: string;
+
+	/**
+	 * The class members with widget decorators.
+	 */
+	members: ThingworxTypescriptWidgetClassMember[];
+}
+
+/**
+ * The primitive type keywords that can be used anywhere.
+ */
+const TypeScriptPrimitiveTypes = [
+	ts.SyntaxKind.StringKeyword,
+	ts.SyntaxKind.NumberKeyword,
+	ts.SyntaxKind.BooleanKeyword,
+	ts.SyntaxKind.AnyKeyword,
+	ts.SyntaxKind.UnknownKeyword
+];
+
+/**
+ * A mapping between typescript type names and valid thingworx base types.
+ */
+ const ThingworxBaseTypes: {[key: string]: string} = {
```

Comment:
It seems to be a bit wrong for this object to contain both uppercase and lowecase the same keys.

Why not just define all of it in lowercase, and when addressing it, address it as `ThingworxBaseTypes[key.toLowerCase()]`

Reply:
The point here is that the keys represent the case-sensitive type names that you can use and their values are the base types they map to. Meaning that you can either use the `STRING` type or the `string` primitive type and it will map to the `STRING` base type, but a custom `StRiNg` type would not. 

The goal here is to allow an alternative to the thingworx base type names (when also properly defined as types) that's easier on the eyes than the all caps names normally used.

Reply:
Hmmm.. I get it. 

You could also use:
```typescript
const ThingworxBaseTypes = {
//                                             ^ let typescript infer the type
    NOTHING: "NOTHING",
    ANYSCALAR: 'ANYSCALAR',
    FIELDNAME: 'FIELDNAME',
    RENDERERWITHFORMAT: 'RENDERERWITHFORMAT',
    STRING: "STRING",
    NUMBER: "NUMBER",
    BOOLEAN: "BOOLEAN",
};

type keys = keyof typeof ThingworxBaseTypes | Lowercase<keyof typeof ThingworxBaseTypes>;
//                    ^ allow for both uppercase (what's declared, and lowercase through https://www.typescriptlang.org/docs/handbook/2/template-literal-types.html#lowercasestringtype)
```

However i'm not exactly sure that this is what you would want, since i feel i'm missing the point completely, and you actually want a runtime check :))

Reply:
Other types with longer names don't just get lowercased; `PERSISTENCEPROVIDERPACKAGENAME` for example gets a camelCase variant `persistenceProviderPackageName` (because `persistenceproviderpackagename` is very unreadable) so a blanket lowercasing wouldn't work well for this object's keys.

Reply:
Yea, i guess there's no getting around it.